annotated.go

```diff
@@ -272,210 +529,357 @@ func (la *lifecycleHookAnnotation) apply(ann *annotated) error {
+			if ctxStructPos < 0 {
+				offset := 0
+				for i := 0; i < len(hookArgs); i++ {
+					if i == ctxPos {
+						hookArgs[i] = reflect.ValueOf(ctx)
+						offset = 1
+						continue
+					}
+					if i-offset >= 0 && i-offset < len(args) {
+						hookArgs[i] = args[i-offset]
+					}
+				}
+			} else {
+				for i := 0; i < origHookFnT.NumIn(); i++ {
+					if i != ctxStructPos {
+						hookArgs[i] = args[i]
+						continue
+					}
+					t := origHookFnT.In(i)
+					v := reflect.New(t).Elem()
+					for j := 1; j < t.NumField(); j++ {
+						if j < ctxPos {
+							v.Field(j).Set(args[i].Field(j))
+						} else if j == ctxPos {
+							v.Field(j).Set(reflect.ValueOf(ctx))
+						} else {
+							v.Field(j).Set(args[i].Field(j - 1))
+						}
+					}
+					hookArgs[i] = v
+				}
+			}
+			hookResults := origHookFn.Call(hookArgs)
+			if len(hookResults) > 0 && hookResults[0].Type() == _typeOfError {
+				err, _ = hookResults[0].Interface().(error)
+			}
 
-func (la *lifecycleHookAnnotation) resolveLifecycleParamField(
-	param reflect.Value,
-	n int,
-) (
-	value reflect.Value,
-) {
-	if param.Kind() == reflect.Struct {
-		if n <= param.NumField() {
-			value = param.FieldByName(fmt.Sprintf("Field%d", n))
+			return
 		}
-	}
+		lc.Append(la.buildHook(hookFn))
+		return
+	})
 
-	return value
+	installerType := reflect.FuncOf(append(resultTypes, paramTypes...), []reflect.Type{_typeOfError}, false)
+	hookInstaller = reflect.MakeFunc(installerType, func(args []reflect.Value) (results []reflect.Value) {
+		results = []reflect.Value{_nilError}
+		var scope *dig.Scope
+		switch la.Type {
+		case _onStartHookType:
+			scope = ann.container.Scope("onStartHookScope")
+		case _onStopHookType:
+			scope = ann.container.Scope("onStopHookScope")
+		}
+		ctor := makeHookScopeCtor(paramTypes, resultTypes, args)
+		if err := scope.Provide(ctor); err != nil {
+			results[0] = reflect.ValueOf(fmt.Errorf("Error providing possible parameters for hook installer: %w", err))
+			return
+		}
+		if err := scope.Invoke(invokeFn.Interface()); err != nil {
+			results[0] = reflect.ValueOf(fmt.Errorf("Error invoking hook installer: %w", err))
+			return
+		}
+		return
+	})
+	return
 }
 
-func (la *lifecycleHookAnnotation) parameters(results ...reflect.Type) (
-	in reflect.Type,
-	argmap func(
-		args []reflect.Value,
-	) (Lifecycle, []reflect.Value),
-) {
-	resultMap := la.resolveMap(results)
+var (
+	_nameTag  = "name"
+	_groupTag = "group"
+)
+
+// makeHookScopeCtor makes a constructor that provides all possible parameters
+// that the lifecycle hook being appended can depend on. It also deduplicates
+// duplicate param and result types, which is possible when using fx.Decorate,
+// and uses values from results for providing the deduplicated types.
+func makeHookScopeCtor(paramTypes []reflect.Type, resultTypes []reflect.Type, args []reflect.Value) interface{} {
+	type key struct {
+		t     reflect.Type
+		name  string
+		group string
+	}
+	seen := map[key]bool{}
```

Comment:
It seems like `seen[key]` is never set to false for any `key` which makes it kind of like a hashset. 

In Go, it is more optimal for you to use an empty struct as the value type in such a case (`map[key]struct{}`). Using this allocates 0 byte for the value as opposed to using bool which allocates 1 byte per value.

Reply:
This is cool. Changed `seen` to `map[key]struct{}`